refactor: column drawing

lewis6991 opened the pull request:
- move some logic out of win_line into specific easy to read
sub-functions.

- remove drawing logic from get_sign_display_info

lewis6991 committed:
refactor: column drawing

- move some logic out of win_line into specific easy to read
  sub-functions.

- remove drawing logic from get_sign_display_info.

zeertzjq commented on src/nvim/screen.c:
```diff
@@ -1948,6 +1948,29 @@ static inline void provider_err_virt_text(linenr_T lnum, char *err)
+    }
+  }
+
+  snprintf((char *)(*extrap), sizeof(*extrap),
```
Using `sizeof(*extrap)` does not work.

zeertzjq replied:
You may need to pass `sizeof(extra)` as an argument.

lewis6991 replied:
Thanks, that saved my time debugging this.

I think I can pass `extra` directly.

zeertzjq commented on src/nvim/screen.c:
```diff
@@ -4569,20 +4549,61 @@ static bool use_cursor_line_sign(win_T *wp, linenr_T lnum)
 }
 
+// Return true if CursorLine highlight is to be used.
+static bool use_cursor_line(win_T *wp, linenr_T lnum, int row, int startrow, int filler_lines)
```
Is it really necessary to split this into a separate function?

zeertzjq commented on src/nvim/screen.c:
```diff
@@ -4569,20 +4549,61 @@ static bool use_cursor_line_sign(win_T *wp, linenr_T lnum)
 }
 
+// Return true if CursorLine highlight is to be used.
+static bool use_cursor_line(win_T *wp, linenr_T lnum, int row, int startrow, int filler_lines)
```
Splitting this into a separate function makes the comments below look strange.

lewis6991 replied:
They look ok to me, or at least not any worse than before.

Can you suggest anything to make them less strange?

zeertzjq replied:
Just do not split this into a separate function.

zeertzjq commented on src/nvim/screen.c:
```diff
@@ -4569,20 +4549,61 @@ static bool use_cursor_line_sign(win_T *wp, linenr_T lnum)
+    if (lnum < wp->w_cursor.lnum) {
+      // Use LineNrAbove
+      return win_hl_attr(wp, HLF_LNA);
+    } else if (lnum > wp->w_cursor.lnum) {
```
```suggestion
    }
    if (lnum > wp->w_cursor.lnum) {
```

zeertzjq commented on src/nvim/screen.c:
```diff
@@ -1948,6 +1948,29 @@ static inline void provider_err_virt_text(linenr_T lnum, char *err)
 }
 
+/// Draw the line number (empty space after wrapping).
+static inline void draw_line_number(win_T *wp, linenr_T lnum, char_u **extrap)
```
Is `inline` really needed here?

lewis6991 replied:
Not sure. 

I only did it because there's no need in creating a stack frame for this, as the change is only to improve readability. With the disclaimer that I'm a C novice, I generally add `inline` to most single use functions, especially when I'm hoisting code out in a refactor like this.

Is there a reason for not adding it? Or just let compiler figure this out.

zeertzjq replied:
Oh, I trying compiling with `RelWithDebInfo` and the `inline` indeed seems to make a difference. This is fine then.

lewis6991 replied:
Interesting...

What difference are you seeing?

zeertzjq commented on src/nvim/screen.c:
```diff
@@ -1948,6 +1948,29 @@ static inline void provider_err_virt_text(linenr_T lnum, char *err)
+    }
+  }
+
+  snprintf((char *)(extra), sizeof(extra),
```
Using `sizeof(extra)` here does not work either. You have to pass `sizeof(extra)` from `win_line()`.

lewis6991 replied:
Sorry of course.

I'm a bit concerned this didn't cause a lot more test failures.

zeertzjq commented on src/nvim/screen.c:
```diff
@@ -4569,20 +4549,62 @@ static bool use_cursor_line_sign(win_T *wp, linenr_T lnum)
+    }
+  }
+
+  if (use_cursor_line(wp, lnum, row, startrow, filler_lines)) {
+    // When 'cursorline' is set highlight the line number of
+    // the current line differently.
+    // When 'cursorlineopt' has "screenline" only highlight
+    // the line number itself.
+    // TODO(vim): Can we use CursorLine instead of CursorLineNr
+    // when CursorLineNr isn't set?
```
I meant these comments here. These comments describe the conditions inside `use_cursor_line()`, so making `use_cursor_line()` a separate function make them look strange. BTW, there comments can now be moved above the `if` line now. They previously appear above the `if` line, and were moved below in <https://github.com/neovim/neovim/commit/bddcbbb5716a005001da3bacb4c1df4ae05e51bc>. Now that there is no longer an `else` branch, they can be moved above again.

lewis6991 replied:
>    // When 'cursorline' is set highlight the line number of
      // the current line differently.

This doesn't look strange.

>    // When 'cursorlineopt' has "screenline" only highlight
      // the line number itself.

This doesn't make any sense to me. The logic has no check for `CULOPT_SCRLINE`

>  // TODO(vim): Can we use CursorLine instead of CursorLineNr
    // when CursorLineNr isn't set?

This is fine and should be placed near `win_hl_attr(wp, HLF_CLN)`.

zeertzjq replied:
"screenline" and "line" cannot be used together. The code checks for `CULOPT_LINE`.

zeertzjq replied:
wait a minute

zeertzjq replied:
"screenline" and "line" cannot be used together, and the code checks for `CULOPT_LINE`. The comment can still be improved though: <https://github.com/vim/vim/pull/9939>.

lewis6991 replied:
I documented the function fully and clarified the line/screenline thing better.

zeertzjq commented on src/nvim/screen.c:
```diff
@@ -4569,20 +4549,62 @@ static bool use_cursor_line_sign(win_T *wp, linenr_T lnum)
          && (wp->w_p_culopt_flags & CULOPT_NBR);
 }
 
+// Return true if CursorLine highlight is to be used.
+static bool use_cursor_line(win_T *wp, linenr_T lnum, int row, int startrow, int filler_lines)
```
Also this checks for whether `CursorLineNr` should be used, not whether `CursorLine` should be used.

zeertzjq committed:
vim-patch:8.2.4557: confusing comment about 'cursorlineopt'

Problem:    Confusing comment about 'cursorlineopt'.
Solution:   Adjust comment.  (closes vim/vim#9939)  Add parenthesis around logical
            OR.
https://github.com/vim/vim/commit/754d2b40369d8fdcf77fc05cc608f86387016bd9

zeertzjq commented:
I renamed `draw_line_number()` to `get_line_number_str()` as it doesn't actually do the drawing.

lewis6991 commented:
All LGTM
